### PR TITLE
bugfix for #563

### PR DESCRIPTION
**Pull request by daisukes**
Use argcomplete's split_line to deal with incomplete quote states.

Signed-off-by: Daisuke Sato <daisukes@cmu.edu>

### CODE REVIEW

File: ros2cli/ros2cli/command/__init__.py [hunk 1]
```diff
@@ -201,7 +200,9 @@ def add_subparsers_on_demand(
         args = argv
         # for completion use the arguments provided by the argcomplete env var
         if _is_completion_requested():
-            args = shlex.split(os.environ['COMP_LINE'])[1:]
+            from argcomplete import split_line
```
Comment (clalancette) [inline ros2cli/ros2cli/command/__init__.py:203]: The big reason not to do this is that `argcomplete` is currently an optional dependency of `ros2cli`, and this makes it required.  Is there another way we can accomplish the same thing?

**Review by clalancette — CHANGES_REQUESTED** [review]

File: ros2cli/ros2cli/command/__init__.py (hunk 1 above)
Comment (daisukes) [reply ros2cli/ros2cli/command/__init__.py:203]: I was wondering if there is a case that we cannot load argcomplete and _is_completion_requested True?

Comment (clalancette) [reply ros2cli/ros2cli/command/__init__.py:203]: I'm honestly not sure.  It looks like the definition of `_is_comletion_requested()` is:

```
def _is_completion_requested():
    return os.environ.get('_ARGCOMPLETE') == '1'
```

I don't know what is expected to set the `_ARGCOMPLETE` environment variable, so I'm not sure what situations get us here.

Comment (daisukes) [reply ros2cli/ros2cli/command/__init__.py:203]: `_ARGCOMPLETE` appears [here](https://github.com/kislyuk/argcomplete/blob/6d33c8e321ffa388c879fec56a80add0181f56b8/argcomplete/shell_integration.py#L44) in argcomplete registration script.

This is called from a ros2cli script, if the script is available meaning argcomplete available I believe,  
https://github.com/ros2/ros2cli/blob/df55a4ed42cdaca1fa103bb1b9807ce2c239055f/ros2cli/completion/ros2-argcomplete.bash#L15-L19

which is invoked when you call setup script
https://github.com/ros2/ros2cli/blob/e5f67d0e0362462644a5530094fac43af84ca938/ros2cli/colcon.pkg#L1-L6

Comment (clalancette) [reply ros2cli/ros2cli/command/__init__.py:203]: Thanks for pointing it out.  This does look OK then.

**Review by clalancette — COMMENTED** [review]
This looks reasonable enough to me.  @claireyywang @mabelzhang can one of you review this, and maybe test out to make sure things still work both with argcomplete installed and with it not installed?

**Review by clalancette — APPROVED** [review]

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13609/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8491/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11325/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13670/)

**Comment by clalancette** [comment]
@daisukes I'd love to get this in.  However, it looks like it is causing some failures in the `test_list_` tests in CI.  Would you mind taking a look?

**Comment by daisukes** [comment]
@clalancette, sorry for the late response. I'm going to look into it!

**Comment by clalancette** [comment]
> @clalancette, sorry for the late response. I'm going to look into it!

Thanks, appreciated!

**Comment by daisukes** [comment]
I've checked the log of the CI and found the error. 
```
FAIL: test_cli.TestROS2TopicCLI.test_topic_endpoint_info_verbose[rmw_fastrtps_cpp]
```
However, this can happen with the previous commit 56db2ba9fcc85ead292b0c3403cf27dba41854e2 of mine.
Also, I tested my patch with the current master, and the patch does not introduce any new error. 

The error is caused by the unexpected output of the `ros2 topic info -v /chatter` command.
It seems like this commit fixed the error.
https://github.com/ros2/ros2cli/commit/2b6930e27713e46c70b95997c511a89bf57cc624

**Comment by claireyywang** [comment]
CI 
Linux [Build Status](https://ci.ros2.org/job/ci_linux/13650/) 
Linux-aarch64 [Build Status](https://ci.ros2.org/job/ci_linux-aarch64/8530/)
macOS [Build Status](https://ci.ros2.org/job/ci_osx/11365/)
Windows [Build Status](https://ci.ros2.org/job/ci_windows/13715/)

**Comment by clalancette** [comment]
> The error is caused by the unexpected output of the `ros2 topic info -v /chatter` command.
> It seems like this commit fixed the error.

Ah, good point.  I've just rebased this branch onto master, and here is another run at CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13652/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8532/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11368/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13717/)

**Comment by clalancette** [comment]
All right, CI is happy.  The PR job is failing, but that doesn't seem to be unique to this job.  @daisukes Thanks for the contribution, and sorry this took so long to get in.  Merging.

**Pull request merged**